File: functions/src/repository/taskRepository.ts
import {
  Firestore,
  Transaction,
  DocumentReference,
} from 'firebase-admin/firestore';
import { Task, TaskState, taskFromFirestore, taskToFirestore } from '../models/task';
import {
  COLLECTION_FLATS,
  COLLECTION_TASKS,
  ERROR_TASK_NOT_FOUND,
} from '../constants/strings';

/**
 * Repository for Task documents under flats/{flatId}/tasks.
 * All Firestore access for tasks goes through this class (Repository pattern).
 */
export class TaskRepository {
  constructor(private readonly db: Firestore) {}

  /** Returns a Firestore document reference for a specific task. */
  private taskRef(flatId: string, taskId: string): DocumentReference {
    return this.db
      .collection(COLLECTION_FLATS)
      .doc(flatId)
      .collection(COLLECTION_TASKS)
      .doc(taskId);
  }

  /** Fetches all tasks for a flat, ordered by ring_index. */
  async getAllTasks(flatId: string): Promise<Task[]> {
    const snapshot = await this.db
      .collection(COLLECTION_FLATS)
      .doc(flatId)
      .collection(COLLECTION_TASKS)
      .orderBy('ring_index')
      .get();

    return snapshot.docs.map((doc) => taskFromFirestore(doc.id, doc.data()));
  }

  /** Fetches all tasks for a flat within a transaction. */
  async getAllTasksInTransaction(
    flatId: string,
    transaction: Transaction,
  ): Promise<Task[]> {
    const collectionRef = this.db
      .collection(COLLECTION_FLATS)
      .doc(flatId)
      .collection(COLLECTION_TASKS)
      .orderBy('ring_index');

    const snapshot = await transaction.get(collectionRef);
    return snapshot.docs.map((doc) => taskFromFirestore(doc.id, doc.data()));
  }

  /** Fetches a single task by ID, throws if not found. */
  async getTask(flatId: string, taskId: string): Promise<Task> {
    const doc = await this.taskRef(flatId, taskId).get();
    if (!doc.exists) {
      throw new Error(`${ERROR_TASK_NOT_FOUND}: ${taskId}`);
    }
    return taskFromFirestore(doc.id, doc.data()!);
  }

  /** Fetches a single task within a transaction. */
  async getTaskInTransaction(
    flatId: string,
    taskId: string,
    transaction: Transaction,
  ): Promise<Task> {
    const doc = await transaction.get(this.taskRef(flatId, taskId));
    if (!doc.exists) {
      throw new Error(`${ERROR_TASK_NOT_FOUND}: ${taskId}`);
    }
    return taskFromFirestore(doc.id, doc.data()!);
  }

  /** Updates specific fields on a task document. */
  async updateTask(
    flatId: string,
    taskId: string,
    updates: Partial<Omit<Task, 'id'>>,
  ): Promise<void> {
    await this.taskRef(flatId, taskId).update(updates);
  }

  /** Updates specific fields on a task document within a transaction. */
  updateTaskInTransaction(
    flatId: string,
    taskId: string,
    updates: Partial<Omit<Task, 'id'>>,
    transaction: Transaction,
  ): void {
    transaction.update(this.taskRef(flatId, taskId), updates as FirebaseFirestore.UpdateData<Task>);
  }

  /** Creates a new task document with a specified ID. */
  async createTask(flatId: string, taskId: string, task: Task): Promise<void> {
    await this.taskRef(flatId, taskId).set(taskToFirestore(task));
  }

  /**
   * Transitions a task from Pending to NotDone (grace period entry).
   * Only acts when the task is still Pending — idempotent if already transitioned.
   */
  async enterGracePeriod(flatId: string, taskId: string): Promise<void> {
    await this.db.runTransaction(async (transaction) => {
      const task = await this.getTaskInTransaction(flatId, taskId, transaction);
      if (task.state !== TaskState.Pending) {
        // Already completed or already in not_done — nothing to do.
        return;
      }
      this.updateTaskInTransaction(
        flatId,
        taskId,
        { state: TaskState.NotDone },
        transaction,
      );
    });
  }
}

File: functions_python/repository/task_repository.py
"""Repository for Task documents under flats/{flatId}/tasks.

All Firestore access for tasks goes through this class (Repository pattern).
"""

from __future__ import annotations

from typing import Any

from constants.strings import (
    COLLECTION_FLATS,
    COLLECTION_TASKS,
    ERROR_TASK_NOT_FOUND,
)
from models.task import Task, TaskState, task_from_firestore, task_to_firestore


class TaskRepository:
    def __init__(self, db: Any) -> None:
        self._db = db

    def _task_ref(self, flat_id: str, task_id: str) -> Any:
        return self._db.collection(COLLECTION_FLATS).document(flat_id).collection(COLLECTION_TASKS).document(task_id)

    def _tasks_collection(self, flat_id: str) -> Any:
        return self._db.collection(COLLECTION_FLATS).document(flat_id).collection(COLLECTION_TASKS)

    def get_all_tasks(self, flat_id: str) -> list[Task]:
        """Fetch all tasks for a flat, ordered by ring_index."""
        snapshot = self._tasks_collection(flat_id).order_by("ring_index").stream()
        return [task_from_firestore(doc.id, doc.to_dict()) for doc in snapshot]

    def get_all_tasks_in_transaction(self, flat_id: str, transaction: Any) -> list[Task]:
        """Fetch all tasks for a flat within a transaction."""
        docs = self._tasks_collection(flat_id).order_by("ring_index").stream(transaction=transaction)
        return [task_from_firestore(doc.id, doc.to_dict()) for doc in docs]

    def get_task(self, flat_id: str, task_id: str) -> Task:
        """Fetch a single task by ID; raise ValueError if not found."""
        doc = self._task_ref(flat_id, task_id).get()
        if not doc.exists:
            raise ValueError(f"{ERROR_TASK_NOT_FOUND}: {task_id}")
        return task_from_firestore(doc.id, doc.to_dict())

    def get_task_in_transaction(self, flat_id: str, task_id: str, transaction: Any) -> Task:
        """Fetch a single task within a transaction; raise ValueError if not found."""
        doc = self._task_ref(flat_id, task_id).get(transaction=transaction)
        if not doc.exists:
            raise ValueError(f"{ERROR_TASK_NOT_FOUND}: {task_id}")
        return task_from_firestore(doc.id, doc.to_dict())

    def update_task(self, flat_id: str, task_id: str, updates: dict[str, Any]) -> None:
        """Update specific fields on a task document."""
        self._task_ref(flat_id, task_id).update(updates)

    def update_task_in_transaction(self, flat_id: str, task_id: str, updates: dict[str, Any], transaction: Any) -> None:
        """Update specific fields on a task document within a transaction."""
        transaction.update(self._task_ref(flat_id, task_id), updates)

    def create_task(self, flat_id: str, task_id: str, task: Task) -> None:
        """Create a new task document with a specified ID."""
        self._task_ref(flat_id, task_id).set(task_to_firestore(task))

    def enter_grace_period(self, flat_id: str, task_id: str) -> None:
        """Transition a task from Pending → NotDone (grace period entry).

        Only acts when the task is still Pending — idempotent if already transitioned.
        """
        from google.cloud.firestore_v1.transaction import transactional

        task_ref = self._task_ref(flat_id, task_id)

        @transactional  # type: ignore[untyped-decorator, unused-ignore]
        def _update_in_tx(tx: Any) -> None:
            doc = task_ref.get(transaction=tx)
            if not doc.exists:
                raise ValueError(f"{ERROR_TASK_NOT_FOUND}: {task_id}")
            if doc.to_dict().get("state") != TaskState.Pending.value:
                # Already completed or already in not_done — nothing to do.
                return
            tx.update(task_ref, {"state": TaskState.NotDone.value})

        _update_in_tx(self._db.transaction())
